### PR TITLE
EVG-20292, EVG-20630: improve idle timeout behavior and logging

### PR DESCRIPTION
**Pull request by Kimchelly**
EVG-20292
EVG-20630

### Description
Idle timeout appears to have been implemented such that it stays alive for the entire task duration, but it can only actually time out a command that runs in pre or main (for example, if a command hits the idle timeout during post, the idle timeout is logged but nothing happens). I didn't want to change this behavior, so instead, I solidified this as the official behavior of idle timeout.

* Run the idle timeout for pre and main only, and stop it when pre and main are over.
* Don't log the "set idle timeout" in blocks where the idle timeout is not relevant (e.g. `post`, `teardown_task`, `teardown_group`, `timeout`).
* As part of conditionally logging the "set idle timeout" log depending on the block, I decided to refactor some of the block naming so it's an actual enum (rather than untyped string) and removed `isTaskCommands` (since you can infer that it's the main task block from the command block type).
* Slightly improve logging for commands so they go to the task logs instead of execution logs (since command failures are important to see) and so they're a little more consistent in the error cases (especially in post group).

### Testing
* Added a bunch of unit tests. Refactors of existing behavior should be covered by the existing agent test suite.
* Tested in staging patches to ensure the idle timeout is as described.

### Documentation
Improved documentation for idle timeout and exec timeout.

### CODE REVIEW

File: agent/agent.go [hunk 1]
```diff
@@ -1092,13 +1099,8 @@ func (a *Agent) runEndTaskSync(ctx context.Context, tc *taskContext, detail *api
 	}
 	defer cancel()
 
-	if err := a.runCommandsInBlock(syncCtx, tc, taskSyncCmds.List(), runCommandsOptions{}, taskSyncBlock); err != nil {
-		tc.logger.Task().Error(message.WrapError(err, message.Fields{
-			"message":    "error running task sync",
-			"total_time": time.Since(start).String(),
```
Comment (malikchaya2) [inline agent/agent.go:1098]: why do we no longer want to log the total time?

File: agent/agent_test.go [hunk 2]
```diff
@@ -484,7 +487,57 @@ post:
...
 		"Finished command 'shell.exec' (step 1 of 1) in block 'post'",
 		"Finished running post-task commands",
-	}, []string{panicLog})
+	}, []string{
+		panicLog,
+		"Set idle timeout for 'shell.exec'",
+	})
+}
+
+func (s *AgentSuite) TestMainTask() {
```
Comment (malikchaya2) [inline agent/agent_test.go:496]: can we maybe either make the task name more descriptive or add a comment explaining what it is that we are asserting with this test?

File: agent/agent_test.go [hunk 3]
```diff
@@ -484,7 +487,57 @@ post:
...
+      script: exit 1
+`
+	s.setupRunTask(projYml)
+
+	s.Error(s.a.runTaskCommands(s.ctx, s.tc))
+	s.NoError(s.tc.logger.Close())
+	checkMockLogs(s.T(), s.mockCommunicator, s.tc.taskConfig.Task.Id, []string{
+		"Running task commands",
+		"Set idle timeout for 'shell.exec'",
+		"Running command 'shell.exec' (step 1 of 1)",
```
Comment (malikchaya2) [inline agent/agent_test.go:536]: Does the failure case not log `"Finished command 'shell.exec' (step 1 of 1)",`?

File: agent/task_context.go [hunk 4]
```diff
@@ -30,7 +30,16 @@ func (tc *taskContext) getCurrentCommand() command.Command {
...
 
-func (tc *taskContext) setCurrentIdleTimeout(cmd command.Command) {
+// setCurrentIdleTimeout sets the idle timeout for the current running command.
+// This timeout only applies to commands running in specific blocks where idle
+// timeout is allowed.
+func (tc *taskContext) setCurrentIdleTimeout(cmd command.Command, block command.BlockType) {
+	switch block {
+	case command.PreBlock, command.SetupGroupBlock, command.SetupTaskBlock, command.MainTaskBlock:
+	default:
+		return
```
Comment (malikchaya2) [inline agent/task_context.go:40]: Should this return an error so that it doesn't result in progammer error down the road?

**Review by malikchaya2 — CHANGES_REQUESTED** [review]

File: agent/agent.go (hunk 1 above)
Comment (Kimchelly) [reply agent/agent.go:1098]: It already logs on L1102 because the return is removed.

File: agent/agent_test.go (hunk 2 above)
Comment (Kimchelly) [reply agent/agent_test.go]: Done, also fixed the other ones already in this file with the same issue.

File: agent/agent_test.go (hunk 3 above)
Comment (Kimchelly) [reply agent/agent_test.go:536]: It doesn't. EVG-20634 is supposed to fully fix this.

File: agent/task_context.go (hunk 4 above)
Comment (Kimchelly) [reply agent/task_context.go]: `setCurrentIdleTimeout` is called even for commands blocks where idle timeout doesn't apply. I changed it to move this block check to the outer function.

File: agent/command.go [hunk 5]
```diff
@@ -179,7 +179,14 @@ func (a *Agent) runCommand(ctx context.Context, tc *taskContext, logger client.L
...
 	tc.setCurrentCommand(cmd)
-	tc.setCurrentIdleTimeout(cmd, options.block)
+	switch options.block {
+	case command.PreBlock, command.SetupGroupBlock, command.SetupTaskBlock, command.MainTaskBlock:
+		// Only set the idle timeout in cases where the idle timeout is actually
+		// respected. In all other blocks, setting the idle timeout should have
+		// no effect.
+		tc.setCurrentIdleTimeout(cmd, options.block)
+	default:
+	}
```
Comment (malikchaya2) [inline agent/command.go:189]: since we don't have a default case, can this be an `if` instead?

**Review by malikchaya2 — CHANGES_REQUESTED** [review]

**Review by malikchaya2 — APPROVED** [review]

**Comment by Kimchelly** [comment]
Re-requesting review since I had to resolve merge conflicts for two different commits into this PR.

**Pull request merged**